### PR TITLE
[2.0.x] ESP32 WiFi interface

### PR DESCRIPTION
**Pull request by grownseed**
### Description

@simon-jouet is doing some amazing work on the ESP32 and has already added support for it in Marlin, so this PR is for taking this a step further and taking advantage of the ESP32's WiFi module.

It adds support for serial over websocket as well as a web UI to control the printer over WiFi. A number of new configuration options have also been added (Web, OTA, ... support). It also uses bonjour so that your printer automatically gets a local domain on your network (http://marlin.local by default).

This is still very much a work in progress, but I'd appreciate any early feedback!

marlin-web-ui

### Benefits

Marlin over WiFi!

### Notes

- the websocket serial implementation has some buffering issues at the moment, seems to have to do with the ESP async server and I'm still looking into that
- the web UI is kept fairly simple and entirely self-contained on purpose, i.e. you shouldn't need an internet connection to run your printer, so no external deps

### CODE REVIEW

**Comment by thinkyhead** [comment]
Good work!

Rebased, squashed, reverted some `platformio.ini` changes, and moved `datadir` under the `esp32` environment. To get the rebased and modified code…
```cpp
git fetch origin
git checkout websocket-serial
git reset --hard origin/websocket-serial
```

**Comment by grownseed** [comment]
@thinkyhead thank you! I'll keep at it and @simon-jouet is sending me a board so that should help.

**Comment by Drachenkaetzchen** [comment]
This sounds pretty cool! I'm right now exploring the possibility to have a standalone ESP32 with attached LCD (some of those pretty cheap 2" TFTs with touch), originally intended to use with OctoPrint - but controlling a printer directly via WiFi is another beast :)

**Comment by psavva** [comment]
How would you connect it all up?
Would you be and to USSR this with ramps for example?

**Comment by psavva** [comment]
How would you connect it all up?
Would you be to use this with Ramps for example?

**Comment by grownseed** [comment]
@psavva please check the details of the work from @simon-jouet here https://github.com/MarlinFirmware/Marlin/issues/6524#issuecomment-406910891, he's building an entire board around the ESP32. Towards the end of that thread is a question similar to yours and some helpful comments from Simon.

**Comment by ModMike** [comment]
How is this different from https://github.com/luc-github/ESP3D-WEBUI? 

Not being cheeky, just trying to understand.

**Comment by vivian-ng** [comment]
@grownseed Has this branch been merged into bugfix-2.0.x? Or should I get a copy from your repository to explore this? Because I do not see the files for web (web.cpp, wifi.cpp, etc.) in bugfix-2.0.x and it seems like an interesting feature to have.

Also, I wonder how much work needs to be done to migrate @luc-github [ESP3D](https://github.com/luc-github/ESP3D) into this branch. This would then offer not just a terminal, but additional web-based controls.

**Comment by thinkyhead** [comment]
> Has this branch been merged into bugfix-2.0.x?

Not yet…
image

**Comment by thinkyhead** [comment]
@grownseed — Rebased and resolved conflicts. To obtain the latest code…
```
git fetch origin
git checkout websocket-serial
git reset --hard origin/websocket-serial
```

**Comment by luc-github** [comment]
@vivian-ng actually I did it in August : https://github.com/luc-github/Marlin/tree/eps32_webui
it need ESPD-WEBUI 2.1
I sent it to Simon as I do not have board to check everything before doing a PR but no news
FYI it is same code as the one used for GRBL_ESP32 (https://github.com/bdring/Grbl_Esp32/tree/WebUI)

**Comment by vivian-ng** [comment]
@luc-github Thank you, I have just cloned your repository (eps32_webui) and will try to compile it and see how it goes.

**Comment by luc-github** [comment]
@vivian-ng feel free to let me know if anything

**Comment by grownseed** [comment]
Sorry for the lack of updates on this, I literally just got my last pieces for my new printer which will be using the ESP32-based controller from @simon-jouet, so I'm seriously hoping to get back to working on this.

@luc-github I don't really have time to look around right now, but it looks like you did some further work on this which is great. Would you be able to provide a quick overview of what your changes are compared to this branch? Thanks!

**Comment by luc-github** [comment]
@grownseed  I did not use your branch as I was doing mine in same time as yours, 
FYR I do not use AsyncWebserver but sync Webserver because we saw some issue when testing on GRBL_ESP32.
Current code is based on latest @simon-jouet code and support most of feature of ESP3D standalone:

It support Captive Portal in AP mode if STA client fail
It support SSDP discovery
It support mDNS discovery
It support Direct SD upload / file browsing / management (time stamp is not implemented yet because I was not sure it was necessary for first part, but code is available)
It support OTA
It support Web Update
It support Terminal commands using WebSocket
It support Admin/User authentication
It support Macro on SPIFFS
It allows to Format SPIFFS
It could support telnet(code ready but not implemented) because I was waiting to see the performance impact first as I do not have a full board but an simple ESP32 + SD card reader

the wifi can also be configured by GCODE, I have used the Reprap:
```
M585: Configure / Display host name

M586: Configure network protocols
Parameters:

Pnn Protocol: 0 = HTTP or HTTPS, 1 = WSocket, 2 = Telnet
Snn 0 = disable this protocol, 1 = enable this protocol
Rnn TCP port number to use for the specified protocol. Ignored unless S = 1. If this parameter is not provided then the default port for that protocol and TLS setting is used.

M587: Set client WiFi host network parameters, or display client network parameters
Parameters:

S"ccc" Network SSID
P"ccc" Network password
Inn.nn.nn.nn (optional) IP address to use when connected to this network. If empty or not specified then an IP address will be acquired via DHCP.
Jnn.nn.nn.nn (optional) Gateway IP address to use when connected to this network.
Knn.nn.nn.nn (optional) Netmask to use when connected to this network

M587 with no parameters display network parameters, but not the remembered password.

M588: Set / Display WiFi mode
Parameters: S0: Wifi Off, S1: STA mode, S2: AP mode , P0 or empty: save only,  P1: apply immediatly
No parameters: Display current and saved wifi mode

M589: Configure access point parameters
Parameters:

S"ccc" The SSID that the WiFi interface should use when it is commanded to run as an access point
P"ccc" The WiFi password
Inn.nn.nn.nn The IP address to use
Cnn The WiFi channel to use (optional)
Note: WPA2 security will be used by default.
```

I may need to update some code to be as up to date as GRBL_ESP32 (to add telnet for example) but my Dev branch is waiting for real stress test.

To sum up I did everything to make it working with ESP3D-WEBUI 2.1

**Comment by grownseed** [comment]
@luc-github that's great to hear, it sounds like you're a lot further into this than I am, so I'm wondering if there's any point taking my branch any further. I might give your branch a stab when I get the chance.

**Comment by luc-github** [comment]
@grownseed sorry I did not wanted to overlap - but most of the work was done already in ESP3D so porting for me was the way to go - the hard time was to work with Marlin SD instead of ESP SD and avoid conflict - but code always have a way ;)

**Comment by vivian-ng** [comment]
@luc-github I managed to compile your eps32-webui branch. I am now waiting for parts to arrive so that I can jury rig a test setup just to check it can control stuff, like move motors, turn on/off heating elements and fans.

**Comment by luc-github** [comment]
@vivian-ng I hope was not difficult - I will try to do some update this week to add telnet support and Serial bluetooth

**Comment by vivian-ng** [comment]
@luc-github It did not compile on Platformio, but I did manage to successfully compile it on Arduino IDE 1.8.5 and flashed it onto a Lolin32 board. I am trying to rig together a simple setup with stepper drivers and mosfets to test whether the board can move the motors and heat up stuff. Waiting for my parts to arrive.

**Comment by marcio-ao** [comment]
We are thinking about adapting this work for the following board for a future printer, since we want Ethernet and WiFi:

https://www.amazon.com/ESP32-GATEWAY-100MBit-Ethernet-smaller-ESP32-EVB/dp/B075C7WVJF

Any thoughts on what it may involve?

**Comment by luc-github** [comment]
without Ethernet the number of pins is already critical - so with ethernet it will be really challenging - but there is always a way I think

**Comment by marcio-ao** [comment]
@luc-github: What I meant is we were going to use the ESP32 board as an add-on alongside an Archim 2.0 board. The idea was to run a webserver on the ESP32, have it handle the web interface and send GCODE over SPI or UART to the Archim 2.0. I assume that in that scenario it wouldn't very many I/O pins?

**Comment by luc-github** [comment]
In case of esp32 as daughter.board you just need serial rx/tx pins if the fw of esp32 act as host
  but in that case why not connect a raspberry pi with octopi ? It bring more feature and the cost compare to the board you mentioned should be close no ?

**Comment by vivian-ng** [comment]
I think having the ESP32 as a built-in module on the board itself is a $5 solution that brings connectivity to any new printer board. In fact, one of my side projects is to design a board based on STM32F407, which Marlin has some support for, with an ESP32 connected to RX/TX to provided web interface via @luc-github ESP3D. But that is paused for now while I work on my ESP32-based board. Once the design has been tested, I will be releasing the design as open source under GPL license. If the design is not released, it means I failed. Haha.

BTW, @luc-github, I have partially hooked up my LOLIN32 board flashed with your [ESP32 web interface branch](https://github.com/luc-github/Marlin/tree/eps32_webui) to a Morpheus board using jumper wires. The motors work, temperature reading seems to work, the only thing I couldn't really test was heating. I think it is because my test power supply (cheap Chinese 12V 5A) couldn't deliver enough juice, causing the rigged setup to reset each time I try to heat the nozzle. But so far, two feedback:
1. Is there a way to enable OLED via I2C on the web interface like in ESP3D v2.1? So that IP address can be seen without having to connect via USB cable.
2. Can the user/admin toggle the need to sign in each time the web interface starts?

Another thing I thought of was EEPROM support. Marlin already has support for I2C EEPROM, so I think it should not be too difficult to add one. But how much work do you think is needed to set aside a portion of flash memory to store the printer settings, and use that to emulate EEPROM? Or is there a way to store printer settings via the web interface, and use that to "configure" the ESP32 each time the printer starts up?

**Comment by luc-github** [comment]
@vivian-ng
1 - Yes, but you you also have SSDP implemented for windows so the device should appear in your Network page:  
image

mDNS is also implemented so you should be able to access using `<hostname>.local`

2 - Just comment `#define ENABLE_AUTHENTICATION` in HAL_ESP32/wificonfig.h

3 - Yes, I see in others HAL `persistent_store_impl.cpp` and `persistent_store_flash.cpp` so we just need to add them and make them to  use the ESP32 EEPROM functions to access to the 4KB nvs reserved for this purpose, it won't be emulation as ESP32 has reserved area for this. 

I think it is better to move discussion to another thread in Marlin issue tracker, or in Simon repository (https://github.com/simon-jouet/Marlin), or even in my repository (https://github.com/luc-github/Marlin) 
Here is a PR and from another person for another project - It would be better not to hack it-  I just answered because I was mentioned, I do not want disrespect @grownseed

**Comment by thinkyhead** [comment]
Rebased, squashed, cleaned up a bit more, and added an ESP32 test for Travis CI. This can be merged pretty soon. To get the latest code use Git Console:
```
git fetch origin
git checkout websocket-serial
git reset --hard origin/websocket-serial
```

**Comment by thinkyhead** [comment]
@grownseed — I've done a lot of cleanup on this to get it down to a good zen level. Please confirm that it all still works for you, and if so then I guess we can go ahead and merge it.
